csm/core/services/usl.py

```diff
@@ -149,273 +146,104 @@ async def _format_lyve_pilot_access_params(
-                    Log.error(f'{reason} --- {e}')
-
-        Log.debug('Create S3 account')
+        iam_user: Optional[Dict] = None
+        iam_user_access_key: Optional[Dict] = None
+        iam_user_access_key_id: Optional[str] = None
+        bucket: Optional[Dict] = None
+
         try:
+            Log.debug('Create S3 account')
             s3_account = await s3_account_service.create_account(
-                s3_account_name, s3_account_email, s3_account_password,
-            )
-            # FIXME `cast()` is unsafe; mypy could not derive type correctly
-            s3_account = cast(Dict, s3_account)
-            access_key = s3_account.get('access_key')
-            secret_key = s3_account.get('secret_key')
+                s3_account_name, s3_account_email, s3_account_password)
             assert s3_account is not None
-        # FIXME There is no way to extract meaningful error codes from ``CsmInternalError``.
-        #   For that reason, we raise internal server errors in all of those cases.
+            s3_credentials = S3Credentials(
+                str(s3_account.get('account_name')),
+                str(s3_account.get('access_key')),
+                str(s3_account.get('secret_key')),
+                '')
+            Log.debug('Create IAM user')
+            iam_user = await s3_iam_users_service.create_user(
+                s3_credentials, iam_user_name, iam_user_password)
+            assert iam_user is not None
+            iam_user_arn = str(iam_user.get('arn'))
+            Log.debug('Create IAM user access key')
+            iam_user_access_key = await s3_access_keys_service.create_access_key(
+                s3_credentials, user_name=iam_user_name)
+            assert iam_user_access_key is not None
+            iam_user_access_key_id = str(iam_user_access_key.get('access_key_id'))
+            Log.debug('Create bucket')
+            bucket = await s3_bucket_service.create_bucket(s3_credentials, bucket_name)
+            assert bucket is not None
+            Log.debug('Tag bucket')
+            await self._tag_lyve_pilot_bucket(s3_credentials, s3_bucket_service, bucket_name)
+            Log.debug('Put bucket policy')
+            await self._set_lyve_pilot_policy(s3_credentials, s3_bucket_service, iam_user_arn, bucket_name)
         except (S3ServiceError, CsmInternalError, Exception) as e:
-            await cleanup(s3_account_service=s3_account_service)
-            reason = 'S3 account creation failed'
+            await self._cleanup(s3_account_service,
+                                s3_iam_users_service,
+                                s3_access_keys_service,
+                                s3_bucket_service,
+                                s3_credentials,
+                                s3_account_name,
+                                iam_user_name,
+                                iam_user_access_key_id,
+                                bucket_name)
```

Comment:
`bucket_name` here is a function argument expected to be non-null. If we provide it to the `_cleanup()` function, it will attempt to delete it although it might not have been created. There is a chance that the `_cleanup()` function would fail then.

The same applies to `iam_user_name`, I think.

Reply:
Fair point. So good you spotted it now! Fixed.